### PR TITLE
[contracts] Add per local weight for function call

### PR DESCRIPTION
**Pull request by agryaznov**
- Limit the number of locals a function can have
- Additional weight for calling a function per local that function has

### CODE REVIEW

**Comment by agryaznov** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_contracts

File: frame/contracts/src/schedule.rs [hunk 1]
```diff
@@ -84,6 +84,9 @@ pub struct Schedule<T: Config> {
 
 	/// The weights for each imported function a contract is allowed to call.
 	pub host_fn_weights: HostFnWeights<T>,
+
+	/// The additional weight for calling a function per local of that function.
+	pub call_per_local_cost: u32,
```
Comment (athei) [inline frame/contracts/src/schedule.rs:89]: Should be part of `InstructionWeights` as `call_per_local` just next to the other call related costs. Also, by putting it here the derived default impl sets it to `0`.

File: frame/contracts/src/benchmarking/mod.rs [hunk 2]
```diff
@@ -2481,6 +2481,22 @@ benchmarks! {
...
+				Instruction::End,
+		]);
+		body::inject_locals(&mut call_body, max_locals);
+		let mut sbox = Sandbox::from(&WasmModule::<T>::from(ModuleDefinition {
+			call_body: Some(call_body),
+			.. Default::default()
+		}));
+	}: {
+		sbox.invoke();
+	}
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:2498]: We need to find out the cost per instruction. Hence we need a component that allows us to do the linear regression over the number of locals. We can't do `r` at the same time as those influence each other.

```suggestion
	// w_per_local = w_bench
	call_per_local {
		let l in 0 .. T::Schedule::get().limits.locals;
		let mut call_body = body::plain(vec![
				Instruction::End,
		]);
		body::inject_locals(&mut call_body, l);
		let mut sbox = Sandbox::from(&WasmModule::<T>::from(ModuleDefinition {
			call_body: Some(call_body),
			.. Default::default()
		}));
	}: {
		sbox.invoke();
	}
```

**Review by athei — CHANGES_REQUESTED** [review]

File: frame/contracts/src/benchmarking/mod.rs [hunk 3]
```diff
@@ -2429,10 +2429,25 @@ benchmarks! {
 		sbox.invoke();
 	}
 
+	// w_per_local = w_bench
+	call_per_local {
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:2433]:
```suggestion
	instr_call_per_local {
```

File: frame/contracts/src/benchmarking/mod.rs [hunk 4]
```diff
@@ -2429,10 +2429,25 @@ benchmarks! {
...
+		let l in 0 .. T::Schedule::get().limits.locals;
+		let mut call_body = body::plain(vec![
+				Instruction::End,
+		]);
+		body::inject_locals(&mut call_body, l);
+		let mut sbox = Sandbox::from(&WasmModule::<T>::from(ModuleDefinition {
```
[the remark this answers is not in the document]
Comment (agryaznov) [reply frame/contracts/src/benchmarking/mod.rs:2439]: but won't it generate just the same function body, once it has just an empty instructions set?

[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:2439]: Have a look at the `instr_call` benchmark. It created an additional function and calls it multiple times. This is what needs to be done here.

**Comment by agryaznov** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_contracts

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2108868

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2108867

**Comment by agryaznov** [comment]
/cmd queue -c bench-bot $ pallet dev pallet_contracts

**Review by xermicus — APPROVED** [review]
LGTM

**Review by athei — APPROVED** [review]

**Comment by athei** [comment]
bot merge

**Pull request merged**